### PR TITLE
Room call banner

### PR DESCRIPTION
**Pull request by toger5**
Closes https://github.com/vector-im/element-web/issues/23453

image

## Checklist

* [x] Tests written for new code (and old code if feasible)
* [ ] Linter and other CI checks pass
* [ ] Sign-off given on the changes (see [CONTRIBUTING.md](https://github.com/matrix-org/matrix-react-sdk/blob/develop/CONTRIBUTING.md))

---
Here's what your changelog entry will look like:

## ✨ Features
 * Room call banner ([\#9378](https://github.com/matrix-org/matrix-react-sdk/pull/9378)). Fixes vector-im/element-web#23453. Contributed by @toger5.

### CODE REVIEW

File: src/components/views/beacon/RoomCallBanner.tsx [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+        case ConnectionState.Connected:
+            [buttonText, buttonKind, onButtonClick] = [
+                _t("Leave"),
+                "danger",
+                disconnect,
+            ];
+            break;
+        case ConnectionState.Disconnecting:
+            [buttonText, buttonKind, onButtonClick] = [
+                _t("Leave"),
```
Comment (toger5) [inline src/components/views/beacon/RoomCallBanner.tsx:112]: The Leave is visible/necessary if the  call is in PiP view.

File: test/components/views/beacon/RoomCallBanner-test.tsx [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+    useMockedCalls,
+} from "../../../test-utils";
+import RoomCallBanner from "../../../../src/components/views/beacon/RoomCallBanner";
+import { CallStore } from "../../../../src/stores/CallStore";
+import { WidgetMessagingStore } from "../../../../src/stores/widgets/WidgetMessagingStore";
+import { MatrixClientPeg } from "../../../../src/MatrixClientPeg";
+import { RoomViewStore } from "../../../../src/stores/RoomViewStore";
+import { ConnectionState } from "../../../../src/models/Call";
+
+describe("<RoomLiveShareWarning />", () => {
```
Comment (robintown) [inline test/components/views/beacon/RoomCallBanner-test.tsx:48]:
```suggestion
describe("<RoomCallBanner />", () => {
```

File: src/components/views/beacon/RoomCallBanner.tsx [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+    call: Call;
+}
+
+const RoomCallBannerInner: React.FC<RoomCallBannerProps> = ({
+    roomId,
+    call,
+}) => {
+    let callEvent: MatrixEvent = null;
+
+    if (!!(call as ElementCall).groupCall) {
```
Comment (robintown) [inline src/components/views/beacon/RoomCallBanner.tsx:47]: I'm confused, why is this condition negated? Would `if (call instanceof ElementCall)` make more sense here?

File: src/components/views/beacon/RoomCallBanner.tsx [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+interface RoomCallBannerProps {
+    roomId: Room["roomId"];
+    call: Call;
+}
+
+const RoomCallBannerInner: React.FC<RoomCallBannerProps> = ({
+    roomId,
+    call,
+}) => {
+    let callEvent: MatrixEvent = null;
```
Comment (robintown) [inline src/components/views/beacon/RoomCallBanner.tsx:45]:
```suggestion
    let callEvent: MatrixEvent | null = null;
```

to make strict mode happy

File: src/components/views/beacon/RoomCallBanner.tsx [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+            action: Action.ViewRoom,
+            room_id: roomId,
+            metricsTrigger: undefined,
+            event_id: callEvent.getId(),
+            scroll_into_view: true,
+            highlighted: true,
+        });
+    };
+
+    let [buttonText, buttonKind, onButtonClick] = [null, null, null, null];
```
Comment (robintown) [inline src/components/views/beacon/RoomCallBanner.tsx:86]: These variables need types, so it may be easier to just split this into 3 separate `let` statements

File: src/components/views/beacon/RoomCallBanner.tsx [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+                "danger",
+                null,
+            ];
+            break;
+    }
+    if (!call) return <Fragment />;
+
+    return (
+        <div
+            className="mx_RoomLiveShareWarning mx_RoomCallBanner"
```
Comment (robintown) [inline src/components/views/beacon/RoomCallBanner.tsx:122]: I'd suggest renaming `mx_RoomLiveShareWarning` to something different if the intent is to share styles between the two banner types

File: src/components/views/beacon/RoomCallBanner.tsx [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+            onClick={onClick}
+        >
+            <div className="mx_RoomCallBanner_text">
+                <span className="mx_RoomCallBanner_label">Video call</span>
```
Comment (robintown) [inline src/components/views/beacon/RoomCallBanner.tsx:126]:
```suggestion
                <span className="mx_RoomCallBanner_label">{ _t("Video call") }</span>
```

File: src/components/views/beacon/RoomCallBanner.tsx [hunk 8]
```diff
@@ -0,0 +1,180 @@
...
+        OwnBeaconStore.instance,
+        OwnBeaconStoreEvent.LivenessChange,
+        () => OwnBeaconStore.instance.getLiveBeaconIds(roomId),
+    );
+
+    if (isMonitoringLiveLocation && liveBeaconIds.length) {
+        return null;
+    }
+
+    // Check if the call is already showing. No banner is needed in this case.
```
Comment (robintown) [inline src/components/views/beacon/RoomCallBanner.tsx:168]: I think we should additionally hide the banner if you're connected to the call, since I don't think the extra 'leave' button is necessary - the picture-in-picture view already serves the purpose of keeping a leave button always available on screen during the call

File: res/css/views/rooms/_RoomCallBanner.pcss [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+            mask-size: 16px;
+            width: 16px;
+            height: 16px;
+            margin-right: 4px;
+            bottom: 2px;
+            mask-image: url("$(res)/img/element-icons/call/video-call.svg");
+        }
+    }
+    .mx_RoomCallBanner_button {
+        left: 16px;
```
Comment (robintown) [inline res/css/views/rooms/_RoomCallBanner.pcss:42]: Is this supposed to be `margin-left`?

File: test/components/views/beacon/RoomCallBanner-test.tsx [hunk 10]
```diff
@@ -0,0 +1,144 @@
...
+    });
+
+    const renderBanner = async (props = {}): Promise<void> => {
+        render(<RoomCallBanner {...defaultProps} {...props} />);
+        await act(() => Promise.resolve()); // Let effects settle
+    };
+
+    it("renders nothing when there is no call", async () => {
+        await renderBanner();
+        const banner = await screen.queryByText("VideoCall");
```
Comment (robintown) [inline test/components/views/beacon/RoomCallBanner-test.tsx:92]:
```suggestion
        const banner = await screen.queryByText("Video call");
```

File: test/components/views/beacon/RoomCallBanner-test.tsx [hunk 11]
```diff
@@ -0,0 +1,144 @@
...
+        afterEach(() => {
+            cleanup(); // Unmount before we do any cleanup that might update the component
+            call.destroy();
+            WidgetMessagingStore.instance.stopMessaging(widget, room.roomId);
+        });
+
+        it("renders if there is a call", async () => {
+            await renderBanner();
+            const videoCallLabel = await screen.findByText("Video call");
+            expect(videoCallLabel.innerHTML).toBe("Video call");
```
Comment (robintown) [inline test/components/views/beacon/RoomCallBanner-test.tsx:120]: I believe that just

```suggestion
            await screen.findByText("Video call");
```

should be sufficient here, since `findByText` will throw an error if it doesn't find the element

File: test/test-utils/call.ts [hunk 12]
```diff
@@ -68,6 +68,10 @@ export class MockedCall extends Call {
         super.participants = value;
     }
 
+    public setConnectionState(value: ConnectionState) {
+        super.connectionState = value;
+    }
```
Comment (robintown) [inline test/test-utils/call.ts:73]: Why was this added?

File: res/css/views/rooms/_RoomCallBanner.pcss [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+.mx_RoomCallBanner {
+    .mx_RoomCallBanner_text {
+        display: flex;
+        flex: 1;
+        align-items: center;
+    }
+    .mx_RoomCallBanner_label {
+        color: $primary-content;
+        font-weight: 600;
+        padding-right: 8px;
```
Comment (kerryarchibald) [inline res/css/views/rooms/_RoomCallBanner.pcss:26]: nitpick: these could use spacing variables - $spacing-8, $spacing-4 etc

File: src/components/views/beacon/RoomCallBanner.tsx [hunk 14]
```diff
@@ -0,0 +1,180 @@
...
+import { Call, ConnectionState, ElementCall } from "../../../models/Call";
+import { useCall } from "../../../hooks/useCall";
+import { RoomViewStore } from "../../../stores/RoomViewStore";
+import { useEventEmitterState } from "../../../hooks/useEventEmitter";
+import {
+    OwnBeaconStore,
+    OwnBeaconStoreEvent,
+} from "../../../stores/OwnBeaconStore";
+import { CallDurationFromEvent } from "../voip/CallDuration";
+import { MockedCall } from "../../../../test/test-utils/call";
```
Comment (dbkr) [inline src/components/views/beacon/RoomCallBanner.tsx:34]: This will be why the CI is unhappy - this will pull test code into the main build and therefore start referencing jest where there is no jest. I see you check to see whether the type of the call is a mockedcall later on, although I'm not quite sure why. Can't the test class provide the thing that the main code is expecting?

File: src/components/views/beacon/RoomCallBanner.tsx [hunk 15]
```diff
@@ -0,0 +1,180 @@
...
+        () => OwnBeaconStore.instance.isMonitoringLiveLocation,
+    );
+
+    const liveBeaconIds = useEventEmitterState(
+        OwnBeaconStore.instance,
+        OwnBeaconStoreEvent.LivenessChange,
+        () => OwnBeaconStore.instance.getLiveBeaconIds(roomId),
+    );
+
+    if (isMonitoringLiveLocation && liveBeaconIds.length) {
```
Comment (kerryarchibald) [inline src/components/views/beacon/RoomCallBanner.tsx:164]: Should add a comment to the corresponding code in the beacon banner to make sure to update this code if the beacon banner display logic changes.

File: src/components/views/beacon/RoomCallBanner.tsx [hunk 16]
```diff
@@ -0,0 +1,180 @@
...
+    call,
+}) => {
+    let callEvent: MatrixEvent = null;
+
+    if (!!(call as ElementCall).groupCall) {
+        callEvent = (call as ElementCall).groupCall;
+    }
+    if (!!(call as MockedCall).event) {
+        callEvent = (call as MockedCall).event;
+    }
```
[the remark this answers is not in the document]
Comment (t3chguy) [reply src/components/views/beacon/RoomCallBanner.tsx:52]: But also our code should not be written to expect Mock structures, the Mocked structures should match the code otherwise your tests won't be covering that realistic objects will do the right thing.

File: src/components/views/rooms/RoomHeader.tsx [hunk 17]
```diff
@@ -733,6 +734,7 @@ export default class RoomHeader extends React.Component<IProps, IState> {
                     { betaPill }
                     { buttons }
                 </div>
+                <RoomCallBanner roomId={this.props.room.roomId} />
```
Comment (kerryarchibald) [inline src/components/views/rooms/RoomHeader.tsx:737]: Could avoid having RoomCallBanner depend on beacon store by doing something like:

```
// banners in order of priority, only one banner rendered at a time
const prioritisedBanners = [
    <RoomLiveShareWarning roomId={this.props.room.roomId} />,
    <RoomCallBanner roomId={this.props.room.roomId} />
].filter(Boolean);
const banner = prioritisedBanners[0] || null;
```

Would keep things simpler in RoomCallBanner, and make this more future proof

**Review by kerryarchibald — APPROVED** [review]
LGTM, left some nitpicks.

File: src/components/views/beacon/RoomCallBanner.tsx (hunk 6 above)
Comment (kerryarchibald) [reply src/components/views/beacon/RoomCallBanner.tsx:122]: my 2c: styles shouldn't be shared, if necessary a reusable component should be abstracted out. In the short term I think it's better to duplicate the styles in `mx_RoomCallBanner`

File: src/components/views/beacon/RoomCallBanner.tsx (hunk 3 above)
Comment (toger5) [reply src/components/views/beacon/RoomCallBanner.tsx:47]: I thought I could just do `call.groupCall` (then I noticed I need to do it case by case because its called `groupCall` and `event`...)

Comment (t3chguy) [reply src/components/views/beacon/RoomCallBanner.tsx:47]: You can if you use type guards as per the link in the other comment :)

File: src/components/views/beacon/RoomCallBanner.tsx (hunk 16 above)
[the remark this answers is not in the document]
Comment (toger5) [reply src/components/views/beacon/RoomCallBanner.tsx:52]: I agree it would be best if it would be called groupCall or event in both structures.

[the remark this answers is not in the document]
Comment (t3chguy) [reply src/components/views/beacon/RoomCallBanner.tsx:52]: `RoomCallBannerProps` says `call` is of type `Call` - which has only `groupCall` 

`MockedCall` extends `Call` and also doesn't have a `event` field

**Review by turt2live — COMMENTED** [review]
looks like this has been reviewed by other people already - no additional comments to add on my side

File: test/test-utils/call.ts (hunk 12 above)
Comment (toger5) [reply test/test-utils/call.ts:73]: For testing the Leave button. I need to easily set the connection state. (`call.connect()` raised issues)

File: src/components/views/rooms/RoomHeader.tsx (hunk 17 above)
Comment (toger5) [reply src/components/views/rooms/RoomHeader.tsx:737]: That makes sense. It would make the tests involve the whole Room Header though.

Comment (toger5) [reply src/components/views/rooms/RoomHeader.tsx:737]: Do you think it is fine leaving it like this in favor of not changing the tests?

File: src/components/views/beacon/RoomCallBanner.tsx (hunk 1 above)
Comment (toger5) [reply src/components/views/beacon/RoomCallBanner.tsx:106]: We decided to just not show the banner if we are in pip mode.

File: test/test-utils/call.ts (hunk 12 above)
Comment (toger5) [reply test/test-utils/call.ts:77]: not needed anymore not

File: src/components/views/beacon/RoomCallBanner.tsx [hunk 18]
```diff
@@ -0,0 +1,134 @@
...
+            action: Action.ViewRoom,
+            room_id: roomId,
+            metricsTrigger: undefined,
+            event_id: callEvent.getId(),
+            scroll_into_view: true,
+            highlighted: true,
+        });
+    }, [callEvent, roomId]);
+
+    if (!call) return <Fragment />;
```
Comment (SimonBrandner) [inline src/components/views/beacon/RoomCallBanner.tsx:70]:
```suggestion
    if (!call) return null;
```

This should work, shouldn't it?

File: src/components/views/beacon/RoomCallBanner.tsx [hunk 19]
```diff
@@ -0,0 +1,134 @@
...
+
+    // Check if the call is already showing. No banner is needed in this case.
+    if (RoomViewStore.instance.isViewingCall()) {
+        return null;
+    }
+
+    // split into outer/inner to avoid watching various parts if there is no call
+    if (call) {
+        // No banner if the call is not connected
+        if (call.connectionState !== ConnectionState.Disconnected) {
```
Comment (SimonBrandner) [inline src/components/views/beacon/RoomCallBanner.tsx:126]: You should use `useConnectingState`. Also, I would personally prefer the if statement to be on one line

File: test/components/views/beacon/RoomCallBanner-test.tsx [hunk 20]
```diff
@@ -0,0 +1,135 @@
...
+    stubClient,
+    useMockedCalls,
+} from "../../../test-utils";
+import RoomCallBanner from "../../../../src/components/views/beacon/RoomCallBanner";
+import { CallStore } from "../../../../src/stores/CallStore";
+import { WidgetMessagingStore } from "../../../../src/stores/widgets/WidgetMessagingStore";
+import { MatrixClientPeg } from "../../../../src/MatrixClientPeg";
+import { RoomViewStore } from "../../../../src/stores/RoomViewStore";
+
+describe("<RoomCallBanner />", () => {
```
Comment (SimonBrandner) [inline test/components/views/beacon/RoomCallBanner-test.tsx:47]: Would be nice to test clicking buttons though due to time, I don't think it is necessary. I would suggest a TODO

**Review by SimonBrandner — APPROVED** [review]
Otherwise LGTM

File: src/components/views/beacon/RoomCallBanner.tsx (hunk 19 above)
Comment (toger5) [reply src/components/views/beacon/RoomCallBanner.tsx:126]: I think useConnectionState does not work in this case since call might be null. (and I cannot setup the hook inside the conditional)

File: src/components/views/beacon/RoomCallBanner.tsx (hunk 18 above)
Comment (toger5) [reply src/components/views/beacon/RoomCallBanner.tsx:70]: This if was redundant anyways (the `...Inner` is only renderd it `call != null`)

File: test/components/views/beacon/RoomCallBanner-test.tsx (hunk 20 above)
Comment (toger5) [reply test/components/views/beacon/RoomCallBanner-test.tsx:47]: added the TODO comment

File: src/components/views/beacon/RoomCallBanner.tsx [hunk 21]
```diff
@@ -172,6 +120,9 @@ const RoomCallBanner: React.FC<Props> = ({ roomId }) => {
 
     // split into outer/inner to avoid watching various parts if there is no call
     if (call) {
+        // No banner if the call is not connected
```
Comment (robintown) [inline src/components/views/beacon/RoomCallBanner.tsx:123]:
```suggestion
        // No banner if the call is not disconnected
```

File: src/components/views/beacon/RoomCallBanner.tsx [hunk 22]
```diff
@@ -172,6 +120,9 @@ const RoomCallBanner: React.FC<Props> = ({ roomId }) => {
...
+        if (call.connectionState !== ConnectionState.Disconnected) {return null;}
```
Comment (robintown) [inline src/components/views/beacon/RoomCallBanner.tsx:124]:
```suggestion
        if (call.connectionState !== ConnectionState.Disconnected) return null;
```

**Review by robintown — APPROVED** [review]

File: src/components/views/beacon/RoomCallBanner.tsx [hunk 23]
```diff
@@ -0,0 +1,131 @@
...
+};
+
+interface Props {
+    roomId: Room["roomId"];
+}
+
+const RoomCallBanner: React.FC<Props> = ({ roomId }) => {
+    const call = useCall(roomId);
+
+    // this section is to check if we have a live location share. If so, we dont show the call banner
```
Comment (dbkr) [inline src/components/views/beacon/RoomCallBanner.tsx:99]: If the comment could mention why, it could be helpful - I'm assuming because they just occupy the same space, in which case this i obviously a bit nonideal & we probably want something a bit more like the toast manager that decides which one takes precedence (but no need to block on that part).

Comment (toger5) [reply src/components/views/beacon/RoomCallBanner.tsx:99]: yes, carry also proposed a better solution than what it currently is. (this makes testing easier though (I think) since it is all inside one component)
